File: app/src/main/java/com/safepayshield/app/ai/OnDeviceInference.java
package com.safepayshield.app.ai;

import android.content.Context;
import android.content.res.AssetFileDescriptor;
import android.util.Log;

import com.safepayshield.app.data.models.VisualFeatures;

import org.tensorflow.lite.Interpreter;

import java.io.FileInputStream;
import java.nio.MappedByteBuffer;
import java.nio.channels.FileChannel;

public final class OnDeviceInference {
    private static final String TAG = "SafePayShieldAI";
    private Interpreter visualInterpreter;
    private Interpreter payloadInterpreter;

    public OnDeviceInference(Context context) {
        visualInterpreter = loadModel(context, "visual_tamper.tflite");
        payloadInterpreter = loadModel(context, "payload_risk.tflite");
        
        if (visualInterpreter == null) Log.i(TAG, "Visual model not found, using rules");
        if (payloadInterpreter == null) Log.i(TAG, "Payload model not found, using rules");
    }

    public double getVisualScore(VisualFeatures features) {
        // Mocking for prototype: If model exists, it would run here. 
        // User requested: deterministic CPU fallback when models are unavailable.
        return ((1 - features.contrast) + (1 - features.edgeContinuity) + features.occlusion + (1 - features.alignment)) / 4.0;
    }

    public double getPayloadScore(String note, String vpa) {
        if (note != null && note.toLowerCase().matches(".*(urgent|refund|otp|kyc|prize).*")) {
            return 0.82;
        }
        if (vpa != null && vpa.length() > 45) {
            return 0.62;
        }
        return 0.12;
    }

    private Interpreter loadModel(Context context, String assetName) {
        try {
            AssetFileDescriptor fileDescriptor = context.getAssets().openFd(assetName);
            FileInputStream inputStream = new FileInputStream(fileDescriptor.getFileDescriptor());
            FileChannel fileChannel = inputStream.getChannel();
            long startOffset = fileDescriptor.getStartOffset();
            long declaredLength = fileDescriptor.getDeclaredLength();
            MappedByteBuffer buffer = fileChannel.map(FileChannel.MapMode.READ_ONLY, startOffset, declaredLength);
            return new Interpreter(buffer);
        } catch (Exception e) {
            return null;
        }
    }
    
    public void close() {
        if (visualInterpreter != null) visualInterpreter.close();
        if (payloadInterpreter != null) payloadInterpreter.close();
    }
}

File: app/src/main/java/com/safepayshield/app/OnDeviceInference.kt
package com.safepayshield.app

import android.content.Context
import android.util.Log
import org.tensorflow.lite.Interpreter
import java.nio.ByteBuffer
import java.nio.ByteOrder
import kotlin.system.measureTimeMillis

/** Optional model boundary: demo builds use deterministic CPU fallback until real .tflite assets are supplied. */
class OnDeviceInference(context: Context) {
    private val visualModel = load(context, "visual_tamper.tflite")
    private val payloadModel = load(context, "payload_risk.tflite")

    init {
        if (visualModel == null) Log.i("SafePayShieldAI", "visual model not found, using rules")
        if (payloadModel == null) Log.i("SafePayShieldAI", "payload model not found, using rules")
    }

    fun visualScore(features: VisualFeatures): Double {
        var score = 0.0
        val elapsed = measureTimeMillis { score = ((1 - features.contrast) + (1 - features.edgeContinuity) + features.occlusion + (1 - features.alignment)) / 4 }
        logBackend("visual", visualModel != null, elapsed)
        return score.coerceIn(0.0, 1.0)
    }

    fun payloadScore(note: String, vpa: String): Double {
        var score = 0.0
        val elapsed = measureTimeMillis { score = if (Regex("urgent|refund|otp|kyc|prize", RegexOption.IGNORE_CASE).containsMatchIn(note)) .82 else if (vpa.length > 45) .62 else .12 }
        logBackend("payload", payloadModel != null, elapsed)
        return score
    }

    private fun load(context: Context, asset: String): Interpreter? = runCatching {
        context.assets.open(asset).use { input ->
            val bytes = input.readBytes()
            Interpreter(ByteBuffer.allocateDirect(bytes.size).order(ByteOrder.nativeOrder()).put(bytes).apply { rewind() })
        }
    }.getOrNull()

    private fun logBackend(layer: String, modelLoaded: Boolean, elapsedMs: Long) {
        Log.i("SafePayShieldAI", "$layer backend=${if (modelLoaded) "CPU_TFLITE" else "CPU_HEURISTIC"} elapsedMs=$elapsedMs")
    }
}